srv/modules/runners/disk_led.py

```diff
@@ -0,0 +1,222 @@
+# -*- coding: utf-8 -*-
+"""
+This runner is used to turn on/off the ident and fault LED of a disk.
+"""
+from __future__ import absolute_import
+
+import logging
+import pprint
+import re
+
+import salt.client
+
+log = logging.getLogger(__name__)
+
+
+def _process(hostname, device_name, led, status):
+    """
+    Turn on/off the specified LED of the disk on the given host.
+    """
+    local_client = salt.client.LocalClient()
+    device_file = '/dev/{}'.format(device_name)
+    status = 'on' if status in [True, 'on'] else 'off'
+
+    # Get the configuration of the command to turn on/off the ident
+    # or fault LED on the target host.
+    result = local_client.cmd(
+        hostname, 'pillar.get', ['disk_led:cmd'], tgt_type='compound')
+    if not result or hostname not in result:
+        raise RuntimeError(
+            'Failed to get "disk_led:cmd" configuration from pillar')
+    cmd_cfg = result[hostname]
+    if not isinstance(cmd_cfg, dict):
+        raise TypeError('Invalid configuration')
+
+    print('Turning {} the {} LED of disk "{}" on host "{}" ...'.format(
+        status, led, device_file, hostname))
+    _cmd_run(hostname, cmd_cfg[led][status].format(device_file=device_file))
+
+
+def _cmd_run(hostname, cmd):
+    """
+    Run the specified command line.
+    .. note:: This function has been introduced for easier unit testing.
+    """
+    local_client = salt.client.LocalClient()
+    result = local_client.cmd(
+        hostname, 'cmd.shell', [cmd], tgt_type='compound', full_return=True)
+    if result and hostname in result:
+        if result[hostname]['retcode'] != 0:
+            raise RuntimeError('Failed to execute command on "{}": {}'.format(
+                hostname, result[hostname]['ret']))
+
+
+def _get_metadata():
+    """
+    Get the OSD metadata. We have to collect this data on the master
+    because it is mostly not possible to collect this data on storage
+    nodes due to missing privileges.
+    :return: Returns the OSD metadata.
+    :rtype: list
+    """
+    __master_opts__ = salt.config.client_config("/etc/salt/master")
+    __master_utils__ = salt.loader.utils(__master_opts__)
+    __salt_master__ = salt.loader.minion_mods(
+        __master_opts__, utils=__master_utils__)
+    master_minion = __salt_master__['master.minion']()
+
+    local_client = salt.client.LocalClient()
+    result = local_client.cmd(
+        master_minion, 'osd.metadata', tgt_type='compound')
+    if not result or master_minion not in result.keys():
+        raise RuntimeError('Failed to get metadata')
+    return result[master_minion]
+
+
+def osd(osd_id, device_type, led, status):
```

Comment:
I was planning on doing all of the osd -> device mapping in ceph-mgr.  I think all we need at this layer is to turn {ident,fault} led {on,off} for a given device on a given host.